### ISSUE TITLE
`--gdn-replay` makes qwen35 generate degenerate looping text

### ISSUE THREAD

**Issue by tmark00** [TheTom/llama-cpp-turboquant#345]
# `--gdn-replay` makes qwen35 generate degenerate looping text

On `fb2cc35ab`, enabling `--gdn-replay` with MTP speculative decoding makes Qwen3.8-27B collapse
into a repeated phrase until `max_tokens`. The flag is EXPERIMENTAL and defaults to off, so nothing
is broken for anyone who leaves it alone — but it currently looks like a large speedup while
actually destroying the output, which makes it easy to adopt by mistake.

## Symptom

```
 Need implement.

 Need implement.

 Need implement.
 ... repeats to max_tokens
```

The first ~900 characters are coherent, then it locks into the loop.

## Why it reads as a speedup

Measured throughput goes **up** by 30%, because looping text is trivial for the draft head to
predict, so acceptance climbs:

|                | acceptance | mean len | tok/s  | rounds/s |
|----------------|-----------:|---------:|-------:|---------:|
| without replay |      0.564 |     2.79 |  51.98 |    45.72 |
| with replay    |      0.808 |     3.90 |  67.52 |    43.61 |

Averages over 4 prompts x 2 passes (second pass with the variant order reversed; results
reproduced to the second decimal). **Rounds per second did not improve** — the actual per-round
work got slightly slower. All of the apparent gain is acceptance.

A single-request example, from the server's own timing lines:

```
without replay:  600 tokens / 11456 ms = 52.37 t/s, graphs reused 201,
                 draft acceptance = 0.60156 (385/640), mean len = 2.80
with replay:     600 tokens /  8788 ms = 68.28 t/s, graphs reused 145,
                 draft acceptance = 0.86990 (448/515), mean len = 3.97
```

## Evidence that this is degeneration, not improvement

**1. Repetitiveness of the generated text.** 600-token generations, reasoning channel, compared to
a `--spec-type none` reference on the same prompt and seed:

| variant                | unique/words | gzip ratio | repeated 3-grams |
|------------------------|-------------:|-----------:|-----------------:|
| `--spec-type none`     |        0.397 |      0.413 |            0.109 |
| MTP, no replay         |        0.260 |      0.378 |            0.228 |
| MTP + `--gdn-replay`   |    **0.143** |  **0.107** |        **0.804** |

A gzip ratio of 0.107 means the text compresses roughly 10:1; 80% of word trigrams repeat; 328
words contain only 47 distinct ones.

**2. Acceptance stops tracking prompt difficulty.** Without replay, acceptance varies with content
as you would expect. With replay, four very different prompts all land in the same narrow band:

| prompt | no replay | with replay |
|---|---:|---:|
| red-black tree in Python | 0.656 | 0.787 |
| how a branch predictor works | 0.466 | 0.777 |
| short story in Polish | 0.494 | 0.837 |
| KV-cache arithmetic derivation | 0.641 | 0.831 |

Uniformly high acceptance across unlike prompts is a degeneracy signature, not a drafting win.

**3. The replay math itself is fine.** `test-gdn-ingredient-replay` passes with
`max_abs = 0.000e+00` on this build. So the fault is not in the ingredient reconstruction but in
how it is wired into the live MTP rollback path.

## Reproduction

```sh
llama-server -m Qwen3.8-27B-Q5_K_M.gguf -ngl 99 -c 131072 -fa on -np 1 \
  -b 2048 -ub 512 --cache-type-k q8_0 --cache-type-v turbo3 \
  --spec-type draft-mtp-adaptive --spec-draft-n-max 4 --spec-draft-n-min-adaptive 2 \
  --reasoning on --reasoning-preserve --gdn-replay
```

Then POST any open-ended prompt with `temperature: 0`, `seed: 42`, `max_tokens: 600`. With
`--reasoning on` the text arrives in `reasoning_content`, not `content` — worth knowing, because a
comparison that only hashes `content` silently compares two empty strings.

## Environment

- RTX 5090 Laptop GPU (sm_120), 24 GB, driver 616.56, Windows 11, MSVC 14.51, CUDA 13.x
- Build `fb2cc35ab`, Release, Ninja, `CMAKE_CUDA_ARCHITECTURES=120-real`, `BUILD_SHARED_LIBS=ON`
- Model: `unsloth/Qwen3.8-27B-GGUF` Q5_K_M

## Not checked

Only this model and this NVIDIA configuration. I did not try KDA layers, qwen35moe, CPU-only, or
non-adaptive `draft-mtp`, and I did not bisect within the replay series
(`b7dce42a2` .. `ac375f654`).

Possibly related: `test-recurrent-state-rollback` fails its third ("dirty ctx") phase on this
build with `dirty-ctx logits mismatch at position 6`. That failure is not caused by this feature —
it reproduces identically on `df7f54729` and `6c7357ae0` — but it is in the same restore path, and
that binary is not registered in ctest, so nothing catches it.

**Comment by giveen** [comment]
@tmark00 thank you for this, the other day I actually started noticing a few issues with this as well, and it slipped out of my mind to look into. I'll add it into my to-do list.

**Comment by giveen** [comment]
Found it. `--gdn-replay` keeps a single recurrent-state checkpoint (`s_ckpt_l`) plus a K-slot ingredient ring, and replays ingredients forward from that checkpoint after a partial rollback (`src/models/delta-net-base.cpp`, `build_recurrent_attn`). The checkpoint-advance logic branches on how many tokens the decode round processed (`n_seq_tokens`) vs. the retained window (`n_rs_seq = --spec-draft-n-max`):

- `n_seq_tokens > n_rs_seq`: checkpoint is correctly advanced to "state exactly `n_rs_seq` tokens before the new end of sequence."
- `n_seq_tokens <= n_rs_seq`: checkpoint is just set to `base_state` (the state at the *start* of the round), on the assumption that it's "still before the window." That's only true when `n_seq_tokens == n_rs_seq` — whenever it's smaller, the checkpoint ends up anchored too close to the sequence end, while ring slots that didn't get overwritten this round still hold stale data from an earlier, unrelated window.

`--spec-draft-n-min-adaptive` legitimately varies draft length round to round, so `n_seq_tokens` regularly comes in below `n_max` — hitting the broken branch on a routine basis, not as an edge case. Fixed-length drafting (plain `draft-mtp`) always takes the first branch, which is almost certainly why this doesn't reproduce there. The next replay then reconstructs state from the wrong baseline — silently, no crash — which matches the symptom (coherent for ~900 chars, then locks into a repeated phrase) much better than a hard failure would. `test-gdn-ingredient-replay` doesn't catch this because it exercises the reconstruction math in isolation, not a sequence of variable-length rounds.

The real fix needs the checkpoint/ring bookkeeping to correctly handle variable-length rounds rather than assuming a fixed per-round cadence. That's more invasive than I want to land blind, so as a first step I'd rather add a guard that rejects (or falls back to full K-snapshots for) the `gdn_replay` + adaptive-drafting combination, so nobody hits silent corruption while the proper fix is worked out. Will follow up with that.

**Comment by TheTom** [comment]
#352 landed on `feature/turboquant-kv-cache` with three fixes to the replay path (rollback depth never consumed, graph reuse ignoring it, conv state never rolled back). Greedy output with the flag on is reported byte-identical to flag-off over 300 tokens on the author's setup. If you can re-run your repro on the current head and post the same metrics, that would confirm it on your side; the multi-sequence case is still a known limitation.

**Comment by tmark00** [comment]
Re-ran the repro on `80007e715` (i.e. with #352 in). Same model, prompt, seed and `max_tokens` as
the original report. **The looping is gone in both drafting modes.** One caveat below.

## Fixed-length drafting: byte-identical, as you reported

`--spec-type draft-mtp --spec-draft-n-max 4`, with and without `--gdn-replay`:

- output SHA256 identical
- `draft acceptance = 0.46819 (390/833), mean len = 2.87` — identical to the digit in both runs

## Adaptive drafting: no longer degenerate, but still not identical

`--spec-type draft-mtp-adaptive --spec-draft-n-max 4 --spec-draft-n-min-adaptive 2`:

- output still differs from the flag-off run
- `draft acceptance` 0.60156 (mean len 2.80) without replay vs 0.50469 (mean len 2.17) with

The direction is the opposite of the pre-fix behaviour, where corruption *inflated* acceptance
because the looping text was trivial to draft. The text itself is healthy now:

| variant | unique/words | gzip ratio | repeated 3-grams |
|---|---:|---:|---:|
| `--spec-type none` (reference) | 0.397 | 0.413 | 0.109 |
| adaptive, no replay | 0.260 | 0.378 | 0.228 |
| **adaptive + replay** | 0.321 | 0.433 | **0.073** |
| fixed, no replay | 0.262 | 0.424 | 0.191 |
| fixed + replay | 0.262 | 0.424 | 0.191 |

Compare the pre-fix numbers for the same configuration: 0.143 / 0.107 / **0.804**. Reading the
adaptive output confirms it — coherent reasoning about duplicate keys in the tree where the old
build was emitting `Need implement.` on repeat.

## The fourth issue is reachable

The one-shot warning added in #352 fires here, 0.11 s into the run, on the first decode:

```
W build_recurrent_attn: gdn_replay: checkpoint taken with n_seq_tokens (1) < n_rs_seq (4);
a rollback deeper than 1 tokens would be incorrect
```

So the `n_seq_tokens < n_rs_seq` branch is not unreachable in practice — `n_seq_tokens` comes in
as 1 against `n_rs_seq` 4.

Being careful about what that does and does not show: the warning is one-shot and fired in **both**
modes, including the fixed-length one whose output is nevertheless byte-identical. So firing alone
is not evidence of damage — it says a deeper rollback would have been wrong, not that one occurred.
I cannot attribute the adaptive divergence to this branch specifically; that would need the
per-round `n_seq_tokens` / rollback-depth trace, which I have not captured. What I can say is that
the condition arises in normal operation, so it is worth closing rather than leaving to the
warning.

## Not measured this time

No throughput or VRAM figures — the machine is off its cooling stand and any timing taken now
would be worthless. Happy to run the interleaved A/B against your 64.0 → 57.1 t/s and −108 MiB
once it is back, if that is useful.

## One build note

`80007e715` does not link on Windows/MSVC: `q8_cache` and `tq_rot_cache` in
`ggml_backend_cuda_context` each nest an identical `struct retired_buf`, and MSVC mangles both as
`retired_buf@<unnamed-tag>@ggml_backend_cuda_context`, so `ggml-cuda.cu.obj` is rejected with
LNK1179 (duplicate COMDAT). I hoisted a single `retired_buf` to class scope locally to get this
build; I will open that separately so it does not clutter this thread.

Environment: RTX 5090 Laptop (sm_120), driver 616.56, Windows 11, MSVC 14.51,
`unsloth/Qwen3.8-27B-GGUF` Q5_K_M, `q8_0`/`turbo3` KV, `-c 131072`, `-np 1`.

**Comment by tmark00** [comment]
Throughput numbers, as offered. Measured on `407f3237b`, laptop on a cooling stand with the vendor
performance profile on, GPU 53-56 C for the whole series and no clock decay.

I used **fixed-length drafting** (`--spec-type draft-mtp --spec-draft-n-max 4`) deliberately: that
is the mode where output is byte-identical with and without the flag, so any time difference is
pure overhead rather than a different token stream. All six runs produced the same output hash and
the same `draft acceptance = 0.46819`, so the two sides did identical work.

Interleaved OFF/ON, three rounds:

| | tok/s | eval time (600 tok) | graphs reused | VRAM |
|---|---:|---:|---:|---:|
| without replay | 50.09 / 51.12 / 50.93 | 11979 / 11737 / 11781 ms | 206 | 23480 MB |
| with replay | 39.21 / 39.74 / 39.33 | 15300 / 15098 / 15254 ms | 50 | 23212 MB |

**−22.3% throughput, −268 MB VRAM.** The cost is roughly double the 11% you measured on the 3090;
the VRAM saving is larger than your 108 MiB, which I assume is the longer context here
(`-c 131072`).

## Where the time goes

The interesting part is `graphs reused`: 206 → 50, while the number of decode rounds is unchanged
(identical output and acceptance). That is ~156 extra graph rebuilds. The extra wall time is
~3.4 s over 600 tokens, so ~22 ms per rebuild — which accounts for essentially the whole
difference. So on this setup the overhead is not the replay work itself; it is lost CUDA graph
reuse.

That is presumably the `can_reuse` change from #352 doing its job: a changed `replay_len` means a
differently shaped reconstruction subtree, so the captured graph is correctly discarded. The
question I cannot answer from outside is whether it has to fire this often. If `replay_len` is 0
on the rounds where nothing was rejected, those rounds should in principle still be able to reuse
a graph, and 50/206 suggests they are not. If that is worth chasing, this configuration looks like
a good testbed — the effect is large and completely reproducible here (three rounds, spread under
0.6 t/s).

## Caveat on the adaptive mode

I did not put a throughput number on `draft-mtp-adaptive`, and I would not trust one: its output
still differs between flag-on and flag-off, so the two runs are not doing the same work and any
tok/s comparison would be measuring the text, not the flag. That is the trap I fell into in the
original report, where corruption made the flag look 30% *faster*.

Environment as before: RTX 5090 Laptop (sm_120), driver 616.56, Windows 11, MSVC 14.51,
`unsloth/Qwen3.8-27B-GGUF` Q5_K_M, `q8_0`/`turbo3` KV, `-c 131072`, `-np 1`, 600 tokens per run,
timings taken from the server's own `eval time` line rather than wall clock.

**Comment by TheTom** [comment]
Closing as fixed by #352. The reporter confirmed the looping is gone on 407f323, fixed-depth output is byte-identical with and without replay, and the remaining adaptive-mode difference is performance rather than the original correctness bug. Thanks for rerunning both modes and posting the throughput data.
